Tmpfile on Windows will often not work, cascades into `PARSE_STACK` error on fix.

Starting from ee3329c8c57, we use `tmpfile` to create temporary files. When running tests, we will get output similar to this on Windows, where I added a `perror` call:

```sh
William@DESKTOP-3H1DSBV MINGW64 ~/Projects/legacy/wla-dx
$ ./run_tests.sh
Building byte_tester...
make ; cp byte_tester ../binaries
make[1]: Entering directory '/home/William/Projects/legacy/wla-dx/byte_tester'
gcc  main.o -o byte_tester
make[1]: Leaving directory '/home/William/Projects/legacy/wla-dx/byte_tester'

########################################################################
WARNING: Valgrind is not installed so we cannot perform memory checks...
########################################################################

Running tests...


rm -f main.o more.o core *~ *.lst linked.rom linked.sym
wla-6502 -i -o main.o main.s
Permission denied
MAIN: Error creating a tmp file for WLA's internal data stream.
make: *** [makefile:15: main.o] Error 1

########
FAILURE!
########
Test "after_test/" of platform "6502/" failed.
```

`perror` returns `Permission denied`. It turns out that for historical reasons, Windows writes tmpfiles to the root directory, which you often don't have permission to write to. [Microsoft docs](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/tmpfile?view=msvc-170#remarks) recommends using `tmpnam` and `fopen`. I've supplied a _partial_ patch here:

```diff
diff --git a/main.c b/main.c
index 782e99a..33348ee 100644
--- a/main.c
+++ b/main.c
@@ -89,6 +89,10 @@ static int s_deletable_structures_max = 0, s_deletable_structures_count = 0;
 extern char *g_sdsctag_name_str, *g_sdsctag_notes_str, *g_sdsctag_author_str;
 #endif
 
+#ifdef WIN32
+  static char * s_tmpfile_name;
+#endif
+
 PROFILE_GLOBALS();
 
 
@@ -526,6 +530,11 @@ static void _procedures_at_exit(void) {
   if (g_file_out_ptr != NULL) {
     fclose(g_file_out_ptr);
     g_file_out_ptr = NULL;
+
+#ifdef WIN32
+    remove(s_tmpfile_name);
+    s_tmpfile_name = NULL;
+#endif
   }
 
   free(g_macro_stack);
@@ -951,7 +960,28 @@ int main(int argc, char *argv[]) {
     return 1;
   }
 
+#ifdef WIN32
+  /* tmpfile uses root of drive on Windows, which is probably not-writable. */
+  s_tmpfile_name = tmpnam(NULL);
+
+  if (s_tmpfile_name == NULL) {
+    fprintf(stderr, "MAIN: Error creating a tmp filename for WLA's internal data stream.\n");
+    return 1;
+  }
+
+  /* Per: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/tempnam-wtempnam-tmpnam-wtmpnam?view=msvc-170&redirectedfrom=MSDN
+  When a file name is prepended with a backslash and no path information,
+  such as \fname21, it indicates that the name is valid for the current working
+  directory. */
+  if(s_tmpfile_name[0] == '\\') {
+    s_tmpfile_name = s_tmpfile_name + 1;
+  }
+
+  g_file_out_ptr = fopen(s_tmpfile_name, "wb");
+  fprintf(stderr, "%s\n", s_tmpfile_name);
+#else
   g_file_out_ptr = tmpfile();
+#endif
   if (g_file_out_ptr == NULL) {
     fprintf(stderr, "MAIN: Error creating a tmp file for WLA's internal data stream.\n");
     return 1;

```

Unfortunately, after applying this patch, the link fails with a `PARSE_STACK` error:

```
$ ./run_tests.sh
Building byte_tester...
make ; cp byte_tester ../binaries
make[1]: Entering directory '/home/William/Projects/legacy/wla-dx/byte_tester'
gcc  main.o -o byte_tester
make[1]: Leaving directory '/home/William/Projects/legacy/wla-dx/byte_tester'

########################################################################
WARNING: Valgrind is not installed so we cannot perform memory checks...
########################################################################

Running tests...


rm -f main.o more.o core *~ *.lst linked.rom linked.sym
wla-6502 -i -o main.o main.s
shm0.
wla-6502 -i -o more.o more.s
s754.
wlalink -v -i -s linkfile linked.rom
more.o: more.s:78: PARSE_STACK: Unresolved reference to "label_2X".
make: *** [makefile:12: all] Error 1

########
FAILURE!
########
Test "after_test/" of platform "6502/" failed.
```

Have any idea on the follow-up error? I've attached my [after_test directory](https://github.com/vhelin/wla-dx/files/12102179/after_test.zip) with assembler output. Patch should be applied against: 45fc096f90b4107e2bdbfcc4f0d07130beac7e7a.

Sorry, I was asleep. You were busy while I was asleep :D!

>What is this shell? I think I need more info about how to set up the same environment that you have...

I'm using [MSYS2](https://www.msys2.org/) to get a `bash` shell _and_ the `mingw-w64` compilers. Looks like you've installed just the `mingw-w64` compilers.

>If you try out the latest sources, does it work now?

Ugh, the documentation _even says_ [tempfiles use](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/tmpfile?view=msvc-170#remarks) `+`!!

> The temporary file is opened in w+b (binary read/write) mode.

I thought it meant `w` _and_ `b`, and forgot the `+` can also be in the middle. I added the `b` locally as well to be extra safe.

Now we get further into the tests, but now we fail at one of the 8080 tests:

```
./run_tests.sh
[30/30] Linking C executable binaries\wla-huc6280.exe
Building byte_tester...
make ; cp byte_tester ../binaries
make[1]: Entering directory '/home/William/Projects/legacy/wla-dx/byte_tester'
gcc  main.o -o byte_tester
make[1]: Leaving directory '/home/William/Projects/legacy/wla-dx/byte_tester'

########################################################################
WARNING: Valgrind is not installed so we cannot perform memory checks...
########################################################################

Running tests...
.............................................................

rm -f core *~ linked.rom linked.sym *.o
wla-8080 -M -o main.o main.s
main.s:138: INCLUDE_FILE: Error creating a tmp file for "includeXYZ3.s"!
main.s:138: ERROR: Couldn't parse ".include".
make: *** [makefile:11: main.o] Error 1

########
FAILURE!
########
Test "makefile_generation/" of platform "8080/" failed.
```

Hmm...

--- a/include.c
+++ b/include.c
@@ -23,6 +23,8 @@ extern FILE *g_file_out_ptr;
 extern struct stringmaptable *g_stringmaptables;
 extern struct string *g_fopen_filenames_first;
 extern int g_is_file_isolated_counter;
+
+extern int create_tmp_file(FILE **);
         
 struct incbin_file_data *g_incbin_file_data_first = NULL, *g_ifd_tmp;
 struct active_file_info *g_active_file_info_first = NULL, *g_active_file_info_last = NULL, *g_active_file_info_tmp = NULL;
@@ -137,7 +139,7 @@ static int _find_file(char *name, FILE **f) {
   /* if we can't find the file, but are only printing makefile rules, silently use an empty file */
   /* a warning might be nice */
   if (g_makefile_rules == YES) {
-    (*f) = tmpfile();
+    create_tmp_file(f);
     if (*f != NULL)
       return SUCCEEDED;
 

--- a/main.c
+++ b/main.c
@@ -85,12 +85,16 @@ struct ext_include_collection g_ext_incdirs;
 static struct structure **s_deletable_structures = NULL;
 static int s_deletable_structures_max = 0, s_deletable_structures_count = 0;
 
+int create_tmp_file(FILE **);
+
 #if defined(Z80)
 extern char *g_sdsctag_name_str, *g_sdsctag_notes_str, *g_sdsctag_author_str;
 #endif
 
 #if defined(WIN32)
-static char *s_tmpfile_name = NULL;
+/* Arbitrary, since TMP_MAX can be rather large. */
+static char * s_tmpfile_names[256];
+static int s_num_tmpfiles = 0;
 #endif
 
 PROFILE_GLOBALS();
@@ -533,10 +537,20 @@ static void _procedures_at_exit(void) {
   }
 
 #if defined(WIN32)
-  if (s_tmpfile_name != NULL) {
-    remove(s_tmpfile_name);
-    s_tmpfile_name = NULL;
+  for(i = 0; i < s_num_tmpfiles; i++) {
+    if (s_tmpfile_names[i] != NULL) {
+      if(s_tmpfile_names[i][0] == '\\')
+        remove(&s_tmpfile_names[i][1]);
+      else
+        remove(s_tmpfile_names[i]);
+      
+      free(s_tmpfile_names[i]);
+      s_tmpfile_names[i] = NULL;
+    }
   }
+
+  s_num_tmpfiles = 0;
+  
 #endif
   
   free(g_macro_stack);
@@ -817,14 +831,33 @@ static int _generate_extra_definitions(void) {
 }
 
 
-static int _create_tmp_file(void) {
-
+int create_tmp_file(FILE ** file_out_ptr) {
 #if defined(WIN32)
-  /* tmpfile uses root of drive on Windows, which is probably not-writable. */
-  s_tmpfile_name = tmpnam(NULL);
+  char * tmpfile_buf;
+  char * tmpfile_name;
 
-  if (s_tmpfile_name == NULL) {
-    fprintf(stderr, "_CREATE_TMP_FILE: Error creating a tmp filename for WLA's internal data stream.\n");
+  *file_out_ptr = NULL;
+
+  if (s_num_tmpfiles >= 255) {
+    fprintf(stderr, "CREATE_TMP_FILE: Maximum number of tmpfiles (256) reached.\n");
+    return FAILED;
+  }
+  
+  /* windows.h MAX_PATH: https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry.
+     Let's not bring in windows.h for a single constant. */
+  tmpfile_buf = malloc(260);
+
+  if (tmpfile_buf == NULL) {
+    fprintf(stderr, "CREATE_TMP_FILE: Error allocating a temporary filename buffer.\n");
+    return FAILED;
+  }
+  s_tmpfile_names[s_num_tmpfiles] = tmpfile_buf;
+
+  /* tmpfile uses root of drive on Windows, which is probably not-writable. */
+  tmpfile_name = tmpnam(s_tmpfile_names[s_num_tmpfiles]);
+
+  if (tmpfile_name == NULL) {
+    fprintf(stderr, "CREATE_TMP_FILE: Error creating a tmp filename.\n");
     return FAILED;
   }
 
@@ -832,18 +865,22 @@ static int _create_tmp_file(void) {
      when a file name is prepended with a backslash and no path information,
      such as \fname21, it indicates that the name is valid for the current working
      directory. */
-  if (s_tmpfile_name[0] == '\\')
-    s_tmpfile_name = s_tmpfile_name + 1;
+  if (tmpfile_name[0] == '\\')
+    tmpfile_name = tmpfile_name + 1;
 
-  g_file_out_ptr = fopen(s_tmpfile_name, "w+");
+  *file_out_ptr = fopen(tmpfile_name, "wb+");
 #else
-  g_file_out_ptr = tmpfile();
+  *file_out_ptr = tmpfile();
 #endif
 
-  if (g_file_out_ptr == NULL) {
-    fprintf(stderr, "_CREATE_TMP_FILE: Error creating a tmp file for WLA's internal data stream.\n");
+  if (*file_out_ptr == NULL) {
+    fprintf(stderr, "CREATE_TMP_FILE: Error creating a tmp file.\n");
     return FAILED;
   }
+
+#if defined(WIN32)
+  s_num_tmpfiles++;
+#endif
 
   return SUCCEEDED;
 }
@@ -994,8 +1031,10 @@ int main(int argc, char *argv[]) {
     return 1;
   }
 
-  if (_create_tmp_file() == FAILED)
+  if (create_tmp_file(&g_file_out_ptr) == FAILED) {
+    fprintf(stderr, "MAIN: Error creating a tmp file for WLA's internal data stream.\n");
     return 1;
+  }
 
   /* small inits */
   if (g_extra_definitions == ON)

--- a/phase_1.c
+++ b/phase_1.c
@@ -146,6 +146,8 @@ extern struct file_name_info *g_file_name_info_first, *g_file_name_info_last, *g
 extern struct incbin_file_data *g_incbin_file_data_first, *g_ifd_tmp;
 extern int g_makefile_rules, g_parsing_function_body, g_force_add_namespace, g_is_file_isolated_counter, g_force_ignore_namespace;
 
+extern int create_tmp_file(FILE **);
+
 static int s_macro_stack_size = 0, s_repeat_stack_size = 0;
 static int s_bank = 0, s_bank_defined = 1, s_line_count_status = ON;
 static int s_block_status = 0, s_block_name_id = 0, s_parse_dstruct_result;
@@ -5869,7 +5871,7 @@ int directive_fopen(void) {
   if (f->f == NULL) {
     if (g_makefile_rules == YES) {
       /* lets just use a tmp file for file operations */
-      f->f = tmpfile();
+      create_tmp_file(&f->f);
       if (f->f == NULL) {
         print_error(ERROR_DIR, "Error creating a tmp file for \"%s\"!\n", f->filename);
         return FAILED;

